tests/integration/__init__.py

```diff
@@ -108,6 +144,25 @@ def cleanup_runtime_config_instance(to_cleanup):
 
 atexit.register(cleanup_runtime_config_instance, RUNTIME_CONFIGS)
 
+_SELECTED_PORTS = set()
+
+
+def get_unused_localhost_port():
+    '''
+    Return a random unused port on localhost
+    '''
+    usock = socket.socket(family=socket.AF_INET, type=socket.SOCK_STREAM)
+    usock.bind(('127.0.0.1', 0))
+    port = usock.getsockname()[1]
+    if port in _SELECTED_PORTS:
+        port = get_unused_localhost_port()
+    usock.close()
+    _SELECTED_PORTS.add(port)
+    return port
+
```

Comment:
You have fixed the possible infinite recursion. Which is great. However, this code is still dangerous. Because you relinquish control of the socket for some duration, anything on the system (which could be totally outside your code) is free to grab and use that port in the interim. This could cause occasional random failures.

A safer approach is to save the socket in `_SELECTED_PORTS`. This can be done by using a dict with the port as the key and the socket as the value. Alternatively, it may be done just by having a flat list of sockets (you can always retrieve the port from the socket if necessary via `getsockname()`).

In this manner, you can avoid recursion altogether. The downside of this is that you will need to use:

```
usock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```

In all places that use these sockets. However, I do believe this is done already in most if not all places that sockets are used.